// @equinor/fusion-cli/lib/build/parts/styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const resolveCliDependency_1 = require("../resolveCliDependency");
const styleLoader = resolveCliDependency_1.default('style-loader');
const cssLoader = resolveCliDependency_1.default('css-loader');
const lessLoader = resolveCliDependency_1.default('less-loader');
exports.default = {
    module: {
        rules: [
            {
                test: /\.css$/,
                use: [
                    styleLoader,
                    {
                        loader: cssLoader,
                        options: {
                            modules: true,
                        },
                    },
                ],
            },
            {
                test: /\.less$/,
                use: [
                    styleLoader,
                    {
                        loader: cssLoader,
                        options: {
                            modules: true,
                        },
                    },
                    {
                        loader: lessLoader,
                    },
                ],
            },
        ],
    },
};
