// @equinor/fusion/lib/intl/Number.js
export const formatNumber = (number, fractionDigits = 2) => {
    const parsedNumber = typeof number === 'number' ? number : parseFloat(number);
    return new Intl.NumberFormat('en-GB', {
        style: 'decimal',
        maximumFractionDigits: fractionDigits,
        minimumFractionDigits: fractionDigits,
    }).format(parsedNumber);
};
export const formatPercentage = (number, fractionDigits = 0) => {
    const parsedNumber = typeof number === 'number' ? number : parseFloat(number);
    return new Intl.NumberFormat('en-GB', {
        style: 'percent',
        maximumFractionDigits: fractionDigits,
        minimumFractionDigits: fractionDigits,
    }).format(parsedNumber);
};
export const formatCurrency = (number, currency = 'NOK', fractionDigits = 2) => {
    const parsedNumber = typeof number === 'number' ? number : parseFloat(number);
    return new Intl.NumberFormat('en-GB', {
        style: 'currency',
        currency,
        maximumFractionDigits: fractionDigits,
        minimumFractionDigits: fractionDigits,
    }).format(parsedNumber);
};
