// @equinor/fusion/lib/http/resourceCollections/BookmarksResourceCollection.js
import BaseResourceCollection from './BaseResourceCollection';
import { combineUrls } from '../../utils/url';
import buildQuery from 'odata-query';
export default class BookmarksResourceCollection extends BaseResourceCollection {
    getBaseUrl() {
        return this.serviceResolver.getBookmarksBaseUrl();
    }
    bookmark(bookmarkId) {
        return combineUrls(this.getBaseUrl(), 'bookmarks', bookmarkId);
    }
    bookmarks() {
        return combineUrls(this.getBaseUrl(), 'persons', 'me', 'bookmarks');
    }
    addBookmark() {
        return combineUrls(this.getBaseUrl(), 'bookmarks');
    }
    deleteBookmark(bookmarkId) {
        return combineUrls(this.getBaseUrl(), 'bookmarks', bookmarkId);
    }
    updateBookmark(bookmarkId) {
        return combineUrls(this.getBaseUrl(), 'bookmarks', bookmarkId);
    }
    applyBookmark(bookmarkId) {
        return combineUrls(this.getBaseUrl(), 'bookmarks', bookmarkId, 'apply');
    }
    addFavouriteBookmark() {
        return combineUrls(this.getBaseUrl(), 'persons', 'me', 'bookmarks', 'favourites');
    }
    deleteFavoriteBookmark(bookmarkId) {
        return combineUrls(this.getBaseUrl(), 'persons', 'me', 'bookmarks', 'favourites', bookmarkId);
    }
    headBookmark(bookmarkId) {
        return combineUrls(this.getBaseUrl(), 'persons', 'me', 'bookmarks', 'favourites', bookmarkId);
    }
    queryBookmarks(filter) {
        const baseUrl = this.bookmarks();
        const odataQuery = buildQuery({ filter });
        return `${baseUrl}${odataQuery}`;
    }
}
