// @equinor/fusion/lib/http/resourceCollections/index.js
import DataProxyResourceCollection from './DataProxyResourceCollection';
import FusionResourceCollection from './FusionResourceCollection';
import ContextResourceCollection from './ContextResourceCollection';
import TasksResourceCollection from './TasksResourceCollection';
import PeopleResourceCollection from './PeopleResourceCollection';
import OrgResourceCollection from './OrgResourceCollection';
import ReportResourceCollection from './ReportResourceCollection';
import PowerBIResourceCollection from './PowerBIResourceCollection';
import NotificationResourceCollection from './NotificationResourceCollection';
import InfoResourceCollection from './InfoResourceCollection';
import UserSettingsResourceCollection from './UserSettingsResoureCollections';
import FusionTasksResourceCollection from './FusionTasksResourceCollection';
import BookmarksResourceCollection from './BookmarksResourceCollection';
export { DataProxyResourceCollection };
export const createResourceCollections = (serviceResolver, options) => ({
    dataProxy: new DataProxyResourceCollection(serviceResolver),
    fusion: new FusionResourceCollection(serviceResolver, options),
    context: new ContextResourceCollection(serviceResolver),
    tasks: new TasksResourceCollection(serviceResolver),
    people: new PeopleResourceCollection(serviceResolver),
    org: new OrgResourceCollection(serviceResolver),
    report: new ReportResourceCollection(serviceResolver),
    powerBI: new PowerBIResourceCollection(serviceResolver),
    notification: new NotificationResourceCollection(serviceResolver),
    info: new InfoResourceCollection(serviceResolver),
    userSettings: new UserSettingsResourceCollection(serviceResolver),
    fusionTasks: new FusionTasksResourceCollection(serviceResolver),
    bookmarks: new BookmarksResourceCollection(serviceResolver),
});
