// @equinor/fusion-web-theme/dist/style-property.js
const camel2kebab = (x) => x.replace(/([a-z0-9])([A-Z])/g, '$1-$2').toLowerCase();
export class StyleProperty {
    constructor(name, value) {
        this.name = name;
        this.value = value;
    }
    static makeCssVar(element, module) {
        return `--fusion-${element}__${module}`;
    }
    static makeCss(obj) {
        return Object.keys(obj)
            .map((x) => `${camel2kebab(x)}: ${obj[x]}`)
            .join(';');
    }
    static makeVariable({ attribute, value }) {
        return `var(${[attribute, value].filter((x) => !!x).join(',')})`;
    }
    static extractVariables(styles) {
        return Object.values(styles).reduce((cur, value) => cur.concat(value.variables), []);
    }
    get style() {
        const { value, attributes } = this;
        return Object.assign(Object.assign({}, value), Object.keys(attributes).reduce((cur, key) => Object.assign(cur, { [key]: StyleProperty.makeVariable(attributes[key]) }), {}));
    }
    get css() {
        const { attributes, style } = this;
        return StyleProperty.makeCss(Object.assign(Object.assign({}, style), Object.keys(attributes).reduce((cur, key) => Object.assign(cur, { [key]: StyleProperty.makeVariable(attributes[key]) }), {})));
    }
    get variables() {
        return this.getVariables();
    }
    getAttributes(include_base) {
        return Object.values(this.attributes).filter((x) => include_base || !x.base);
    }
    getVariables(include_base) {
        return this.getAttributes(include_base).map(({ attribute, value }) => [attribute, value].join(':'));
    }
    toString() {
        return this.css;
    }
}
export default StyleProperty;
