// @equinor/fusion/lib/http/apiClients/index.js
import DataProxyClient from './DataProxyClient';
import FusionClient from './FusionClient';
import ContextClient from './ContextClient';
import TasksClient from './TasksClient';
import PeopleClient from './PeopleClient';
import OrgClient from './OrgClient';
import ReportClient from './ReportClient';
import PowerBIClient from './PowerBIClient';
import NotificationClient from './NotificationClient';
import InfoClient from './InfoClient';
import UserSettingsClient from './UserSettingsClient';
import FusionTasksClient from './FusionTasksClient';
export const createApiClients = (httpClient, resources, serviceResolver) => ({
    dataProxy: new DataProxyClient(httpClient, resources, serviceResolver),
    fusion: new FusionClient(httpClient, resources, serviceResolver),
    context: new ContextClient(httpClient, resources, serviceResolver),
    tasks: new TasksClient(httpClient, resources, serviceResolver),
    people: new PeopleClient(httpClient, resources, serviceResolver),
    org: new OrgClient(httpClient, resources, serviceResolver),
    report: new ReportClient(httpClient, resources, serviceResolver),
    powerBI: new PowerBIClient(httpClient, resources, serviceResolver),
    notification: new NotificationClient(httpClient, resources, serviceResolver),
    info: new InfoClient(httpClient, resources, serviceResolver),
    userSettings: new UserSettingsClient(httpClient, resources, serviceResolver),
    fusionTasks: new FusionTasksClient(httpClient, resources, serviceResolver),
});
