// @equinor/fusion/lib/http/apiClients/UserSettingsClient.js
import BaseApiClient from './BaseApiClient';
export class UserSettingsClient extends BaseApiClient {
    getBaseUrl() {
        return this.serviceResolver.getFusionBaseUrl();
    }
    async getAppUserSettings(appKey) {
        const url = this.resourceCollections.userSettings.appUserSettings(appKey);
        return await this.httpClient.getAsync(url);
    }
    async updateAppUserSettings(appKey, userSettings) {
        const url = this.resourceCollections.userSettings.appUserSettings(appKey);
        return await this.httpClient.putAsync(url, userSettings);
    }
    async deleteAllAppUserSettings(appKey) {
        const url = this.resourceCollections.userSettings.appUserSettings(appKey);
        return await this.httpClient.deleteAsync(url);
    }
    async getFusionUserSettings() {
        const url = this.resourceCollections.userSettings.fusionUserSettings();
        return await this.httpClient.getAsync(url);
    }
    async updateFusionUserSettings(userSettings) {
        const url = this.resourceCollections.userSettings.fusionUserSettings();
        return await this.httpClient.putAsync(url, userSettings);
    }
    async deleteAllFusionUserSettings() {
        const url = this.resourceCollections.userSettings.fusionUserSettings();
        return await this.httpClient.deleteAsync(url);
    }
}
export default UserSettingsClient;
