// @equinor/fusion-web-theme/dist/styles/colors.js
import { tokens } from '@equinor/eds-tokens';
import StyleProperty from '../style-property';
export class ColorStyleProperty extends StyleProperty {
    get attributes() {
        const { name, value: { rgba }, } = this;
        const attribute = StyleProperty.makeCssVar('color', name);
        return {
            color: { attribute, value: rgba },
        };
    }
    get style() {
        return {
            color: this.value.rgba,
        };
    }
}
const createStyleProperty = (obj, name) => Object.keys(obj).reduce((cur, key) => Object.assign(cur, { [key]: new ColorStyleProperty(`${name}-${key}`, obj[key]) }), {});
export const colors = {
    infographic: createStyleProperty(tokens.colors.infographic, 'infographic'),
    interactive: createStyleProperty(tokens.colors.interactive, 'interactive'),
    logo: createStyleProperty(tokens.colors.logo, 'logo'),
    text: createStyleProperty(tokens.colors.text, 'text'),
    ui: createStyleProperty(tokens.colors.ui, 'ui'),
};
export const colorVariables = Object.values(colors).reduce((cur, value) => cur.concat(ColorStyleProperty.extractVariables(value)), []);
export default colors;
